datafusion/core/src/physical_plan/unnest.rs

```diff
@@ -0,0 +1,305 @@
+                    );
+                    other
+                }
+            })
+    }
+}
+
+fn build_batch(
+    batch: &RecordBatch,
+    schema: &SchemaRef,
+    column: &Column,
+) -> Result<RecordBatch> {
+    let list_array = column.evaluate(batch)?.into_array(batch.num_rows());
+    match list_array.data_type() {
+        arrow::datatypes::DataType::List(_) => {
+            let list_array = list_array.as_any().downcast_ref::<ListArray>().unwrap();
+            unnest_batch(batch, schema, column, &list_array)
+        }
+        arrow::datatypes::DataType::LargeList(_) => {
+            let list_array = list_array
+                .as_any()
+                .downcast_ref::<LargeListArray>()
+                .unwrap();
+            unnest_batch(batch, schema, column, &list_array)
+        }
+        arrow::datatypes::DataType::FixedSizeList(_, _) => {
+            let list_array = list_array
+                .as_any()
+                .downcast_ref::<FixedSizeListArray>()
+                .unwrap();
+            unnest_batch(batch, schema, column, list_array)
+        }
+        _ => {
+            return Err(DataFusionError::Execution(format!(
+                "Invalid unnest column {column}"
+            )));
+        }
+    }
+}
+
+fn unnest_batch<T>(
+    batch: &RecordBatch,
+    schema: &SchemaRef,
+    column: &Column,
+    list_array: &T,
+) -> Result<RecordBatch>
+where
+    T: ArrayAccessor<Item = ArrayRef>,
+{
+    let mut batches = Vec::new();
+    let mut num_rows = 0;
+
+    for row in 0..batch.num_rows() {
+        let arrays = batch
+            .columns()
+            .iter()
+            .enumerate()
+            .map(|(col_idx, arr)| {
+                if col_idx == column.index() {
+                    // Unnest the value at the given row.
+                    if list_array.value(row).is_empty() {
+                        // If nested array is empty add an array with 1 null.
+                        Ok(new_null_array(list_array.value(row).data_type(), 1))
+                    } else {
+                        Ok(list_array.value(row))
+                    }
+                } else {
+                    // Number of elements to duplicate, use max(1) to handle null.
+                    let nested_len = list_array.value(row).len().max(1);
+                    // Duplicate rows for each value in the nested array.
+                    if arr.is_null(row) {
+                        Ok(new_null_array(arr.data_type(), nested_len))
+                    } else {
+                        let scalar = ScalarValue::try_from_array(arr, row)?;
+                        Ok(scalar.to_array_of_size(nested_len))
+                    }
+                }
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let rb = RecordBatch::try_new(schema.clone(), arrays.to_vec())?;
+        num_rows += rb.num_rows();
+        batches.push(rb);
+    }
+
+    concat_batches(schema, &batches, num_rows).map_err(Into::into)
```

Comment:
this also results in another copy -- aka it is an optimization for the future

Reply:
Added note in #5198